File: 113-path-sum-ii/113-path-sum-ii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> List[List[int]]:
        if not root:
            return []
        path = [root.val]
        final = []
        def find_target(root,targetSum):
            if not root.left and not root.right:
                if targetSum != 0:
                    return False
                elif targetSum == 0:
                    final.append(path[:])
                    return True
#            path = path.append(root.val)
            for nxt in [root.left, root.right]:
                if not nxt:
                    continue
                path.append(nxt.val)
                find_target(nxt,targetSum - nxt.val)
                path.pop()
        find_target(root,targetSum-root.val)
        return final

File: 113-path-sum-ii/113-path-sum-ii.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {

    public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
        List<List<Integer>> pathsList = new ArrayList<List<Integer>>();
        List<Integer> pathNodes = new ArrayList<Integer>();
        recurseTree(root, targetSum, pathNodes, pathsList);
        return pathsList;
        
        
        
    }
    
    private void recurseTree(TreeNode node, int remainingSum, List<Integer> pathNodes, List<List<Integer>> pathsList){
        if(node == null)return;
        pathNodes.add(node.val);
        //check if the current node is a leaf and also if it equals our remaining sum. If it does, add the path to our list of paths.
        if(remainingSum == node.val && node.left == null && node.right == null){
            pathsList.add(new ArrayList<>(pathNodes));
        }else{
            recurseTree(node.left, remainingSum - node.val, pathNodes, pathsList);
            recurseTree(node.right, remainingSum - node.val, pathNodes, pathsList);
        }
        pathNodes.remove(pathNodes.size() - 1);
    }
}